File: inst/tinytest/test_geom.R

r <- rast(ncol = 20, nrow = 10, xmin=-10, xmax=10, ymin=-5, ymax=6)
values(r) <- ncell(r):1
r <- c(r, r/2, r*2)
	
# compare geometries
expect_equal(res(r), c(1.0, 1.1)) 
expect_equal(dim(r), c(10, 20, 3)) 
expect_equivalent(as.vector(ext(r)), c(-10, 10, -5, 6)) 
expect_equal(xyFromCell(r, 10),  cbind(x=-0.5, y=5.45))
expect_equal(xyFromCell(r, 1), cbind(x=-9.5, y=5.45))

expect_equal(xyFromCell(r, ncell(r)), cbind(x=9.5, y=-4.45))
expect_equal(as.vector(is.na(xyFromCell(r, 0))), c(TRUE, TRUE))
expect_equal(as.vector(is.na(xyFromCell(r, ncell(r)+1))), c(TRUE, TRUE))

expect_equivalent(r[10], data.frame(cbind(191, 95.5, 382)))
expect_equivalent(r[2,], data.frame(matrix(c(180,179,178,177,176,175,174,173,172,171,170,169,168,167,166,165,164,163,162,161,90,89.5,89,88.5,88,87.5,87,86.5,86,85.5,85,84.5,84,83.5,83,82.5,82,81.5,81,80.5,360,358,356,354,352,350,348,346,344,342,340,338,336,334,332,330,328,326,324,322), ncol=3)))
expect_equivalent(r[,2], data.frame(matrix(c(199,179,159,139,119,99,79,59,39,19,99.5,89.5,79.5,69.5,59.5,49.5,39.5,29.5,19.5,9.5,398,358,318,278,238,198,158,118,78,38), ncol=3)))
expect_equivalent(r[3:4, 2:3], data.frame(matrix(c(159,158,139,138,79.5,79,69.5,69,318,316,278,276),ncol=3)))
expect_equivalent(r[[1]][5,6][1], data.frame(lyr.1=115))

File: python/tests/test_geom.py
"""
Tests ported from inst/tinytest/test_geom.R

Covers: resolution, dimensions, extent, and cell-to-xy conversion.
"""
import math
import numpy as np
import pytest
import geospat as pt
from geospat.rast import rast
from geospat.values import set_values
from geospat.cells import xy_from_cell


def make_r3():
    """
    Three-layer 10×20 raster with values 200..1 in the first layer,
    and r/2 and r*2 as the second and third layers.
    """
    r = rast(nrows=10, ncols=20, xmin=-10, xmax=10, ymin=-5, ymax=6)
    r = set_values(r, list(range(200, 0, -1)))
    r2 = r * 2
    r_half = r / 2      # Python operator via arith.py
    # Note: r2 is r*2, r_half is r/2 — but we need r, r/2, r*2 as layers
    opt = pt.SpatOptions()
    out = r.deepcopy()
    out.addSource(r_half, True, opt)
    out.addSource(r2, True, opt)
    return out


def test_resolution():
    r = make_r3()
    assert r.xres() == pytest.approx(1.0)
    assert r.yres() == pytest.approx(1.1)


def test_dimensions():
    r = make_r3()
    assert r.nrow() == 10
    assert r.ncol() == 20
    assert r.nlyr() == 3


def test_extent():
    r = make_r3()
    assert r.xmin() == pytest.approx(-10)
    assert r.xmax() == pytest.approx(10)
    assert r.ymin() == pytest.approx(-5)
    assert r.ymax() == pytest.approx(6)


def test_xy_from_cell_10():
    """0-based cell 9 (same location as R / terra cell 10) → x=-0.5, y=5.45."""
    r = make_r3()
    xy = xy_from_cell(r, 9)
    assert xy[0][0] == pytest.approx(-0.5)
    assert xy[0][1] == pytest.approx(5.45)


def test_xy_from_cell_1():
    """First cell (0-based index 0, top-left centre) → x=-9.5, y=5.45."""
    r = make_r3()
    xy = xy_from_cell(r, 0)
    assert xy[0][0] == pytest.approx(-9.5)
    assert xy[0][1] == pytest.approx(5.45)


def test_xy_from_cell_last():
    """Last cell → x=9.5, y=-4.45."""
    r = make_r3()
    ncell = r.ncell()
    xy = xy_from_cell(r, ncell - 1)
    assert xy[0][0] == pytest.approx(9.5)
    assert xy[0][1] == pytest.approx(-4.45, rel=1e-4)


def test_xy_from_cell_out_of_bounds():
    """Cell -1 or ncell → NaN coords (or out-of-range indicator)."""
    r = make_r3()
    ncell = r.ncell()
    for bad_cell in (-1, ncell):
        xy = xy_from_cell(r, bad_cell)
        # Out-of-range 0-based indices; check coords are NaN or extreme.
        x_coord = float(xy[0][0])
        y_coord = float(xy[0][1])
        is_invalid = (
            math.isnan(x_coord) or math.isnan(y_coord)
            or abs(x_coord) > 1e15 or abs(y_coord) > 1e15
        )
        assert is_invalid, (
            f"Expected invalid coords for cell {bad_cell}, got ({x_coord}, {y_coord})"
        )


